pkg/metrics/metrics.go

```diff
@@ -0,0 +1,90 @@
+package metrics
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	// Default values which are to be used as labels
+	// for prometheus metrics.
+	ERROR      = "opsrc_error"
+	OPSRC_NAME = "opsrc_name"
+
+	// Path, Port and Host where custom metrics would be exposed.
+	METRICSPATH = "/metrics"
+	METRICSPORT = "8383"
+	METRICSHOST = "0.0.0.0"
+
+	// Error message to be returned when quay.io is down.
+	UNREACHABLE_ERROR = "Quay.io is unavailabe"
+
+	// Error message to be returned when host address is wrong.
+	NO_HOST_EXSISTS_ERROR = "No such host exsists. Check the quay endpoint"
+)
+
+// Metric variables that report error based on the operator-source
+// which fails to reconcile.
+var (
+	operatorSourceError = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "operator_source_failure_count",
+			Help: "Monotonic count of times operator-source could not download packages",
+		},
+		[]string{OPSRC_NAME, ERROR},
+	)
+	// List of custom metrics which are to be collected.
+	metricsList = []prometheus.Collector{
+		operatorSourceError,
+	}
+)
+
+// registerMetrics registers the metrics with prometheus.
+func registerMetrics() error {
+	for _, metric := range metricsList {
+		err := prometheus.Register(metric)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RecordMetrics starts the server, records the custom metrics and exposes
+// the metrics at the specified endpoint.
+func RecordMetrics() {
+	// Register metrics for the operator with the prometheus.
+	registerMetrics()
+
+	// Start the server and expose the registered metrics.
+	http.Handle(METRICSPATH, promhttp.Handler())
+	port := ":" + METRICSPORT
+	go http.ListenAndServe(port, nil)
```

Comment:
This is registering and serving prom from the global muxer - I assume there's a health check also on it somewhere?

That would mean the health check is exposed on two ports, which probably won't cause an error but is kind of weird.

Reply:
Yes, these are the same port and path values which SDK uses to expose metrics and health check occurs here too. 
Shall I change, or is it fine to use it from a global mux?

Reply:
Ah - if they're both configured on the same port that's fine. I should've checked 😄 I assumed they were configured on different ports.